curvefs/src/mds/space/volume_space.cpp

```diff
@@ -487,6 +650,235 @@ SpaceErrCode VolumeSpace::ExtendVolume() {
     return SpaceOk;
 }
 
+void VolumeSpace::Run() {
+    calThread_ = std::thread([&] {
+        while (sleeper_.wait_for(std::chrono::seconds(calcIntervalSec_))) {
+            CalBlockGroupAvailableForDeAllocate();
+        }
+    });
+    LOG(INFO) << "VolumeSpace start background, fsid=" << fsId_;
+}
+
+void VolumeSpace::Stop() {
+    LOG(INFO) << "VolumeSpace stopping, fsid=" << fsId_;
+
+    sleeper_.interrupt();
+    if (calThread_.joinable()) {
+        calThread_.join();
+    }
+
+    LOG(INFO) << "VolumeSpace stopped, fsid=" << fsId_;
+}
+
+bool VolumeSpace::UpdateDeallocatableBlockGroup(
+    uint32_t metaserverId, uint32_t metaserverNum,
+    const DeallocatableBlockGroupVec &groups,
+    const BlockGroupDeallcateStatusMap &stats, uint64_t *issue) {
+    uint32_t current = metaserverNum_.load(std::memory_order_acquire);
+    metaserverNum_.compare_exchange_strong(current, metaserverNum,
+                                           std::memory_order_acq_rel);
+
+    UpdateBlockGroupDeallocatableSpace(metaserverId, groups);
+
+    UpdateDeallocatingBlockGroup(metaserverId, stats);
+
+    return SelectBlockGroupForDeAllocate(metaserverId, issue);
+}
+
+void VolumeSpace::UpdateBlockGroupDeallocatableSpace(
+    uint32_t metaserverId, const DeallocatableBlockGroupVec &groups) {
+    LockGuard statlk(statmtx_);
+    VLOG(6) << "VolumeSpace update from metaserver:" << metaserverId
+            << ", fsId=" << fsId_ << ", groups size=" << groups.size();
+
+    // update summary_ with latest groups
+    std::unordered_map<uint64_t, uint64_t> reportGroups;
+    for (auto &group : groups) {
+        auto offset = group.blockgroupoffset();
+        auto deallocatableSize = group.deallocatablesize();
+        reportGroups[offset] = deallocatableSize;
+
+        auto iter = summary_.find(offset);
+        if (iter == summary_.end()) {
+            summary_.emplace(offset, deallocatableSize);
+        } else {
+            iter->second += group.deallocatablesize();
+        }
+        VLOG(6) << "VolumeSpace update summary, fsId=" << fsId_
+                << ", blockGroupOffset=" << offset
+                << ", deallocatableSize=" << group.deallocatablesize();
+    }
+
+    // remove groups from last round of reporting and record latest in
+    // lastUpdate_
+    auto lastUpdateIter = lastUpdate_.find(metaserverId);
+    if (lastUpdateIter == lastUpdate_.end()) {
+        lastUpdate_[metaserverId] = groups;
+    } else {
+        for (auto &group : lastUpdateIter->second) {
```

Comment:
The update logic of this function is:
1. all the deallocatablesizes reported by the metaserver are added to the summary_
2. the last report needs to be removed from the summary_, because the metaserver reports the full amount of data each time

so, 10 in lastUpdater should remove deallocatable size from summary_, and 10 new do not need.